index.rst

```diff
@@ -1,29 +1,481 @@
+Introducing a locking mechanism would also permit mixing different schemes within a single workflow.
+
+Dependent DAG Jobs
+------------------
+
+Instead of making all aggregation the responsibility of a single ``finalJob``, we could add multiple aggregation jobs throughout the workflow DAG.
+These jobs would have to be placed in the DAG such that they do not run in parallel with each other, and they'd need depend on regular jobs to space them out temporally (e.g. we could have one aggregation job for each task).
+In this scheme each aggregation job would look only for the quanta it depends on in the workflow DAG, and then exit.
+A regular ``finalJob`` would also be included to take care of any aggregation not done by the previous jobs, as well as constructing and ingesting the final provenance graph.
+
+A major disadvantage of this scheme is that each aggregation job would only get executed if all of its upstream jobs actually succeeded; running a job regardless of upstream success or failure as in ``finalJob`` may or may not be possible for different WMSs, and it would at the very least require additional BPS work.
+An early failure could thus prevent all aggregation jobs from ever becoming unblocked, leaving too much work for ``finalJob``.
+While this would still be an improvement on the status quo (since quanta usually succeed, and failures also block outputs from being produced), it would likely not fully solve the problem of slow ``finalJob`` performance.
+
+Single Monitor Job
+------------------
+
+Because we envision an aggregation tool that could be configured to alternatively sleep and poll for newly-completed quanta (via metadata and log existence checks, targeting only the unblocked quanta), we could delegate all aggregation work to a single workflow DAG job that is launched very early in the processing (e.g. after ``pipetaskInit``).
+For fault tolerance, this job would need to be configured to automatically retry indefinitely (unless explicitly killed with ``bps cancel``), and it would complete successfully only when all predicted quanta are aggregated (whether successful or not).
+We would also need to find a way to ensure it is scheduled for execution early rather than late, since it would "tie" with essentially all regular jobs in any purely topological ordering of the DAG.
+
+A regular ``finalJob`` would also be included, with a dependency on the long-lived aggregation job as well as the regular jobs, but in the usual case it would only be responsible for constructing and ingesting the provenance graph from a fully-populated aggregation graph.
+When a workflow is canceled (but ``finalJob`` is permitted to run), the ``finalJob`` would take over responsibility for aggregating the outputs of any remaining quanta.
+
+The main disadvantage of this scheme is that using existence checks to poll for metadata and log datasets for completed results in more load on the storage system.
```

Comment:
Which is at least less load than checking all quantum outputs as we do now (but if we are setting up a monitor job that polls we are better off setting up a monitor job that receives kafka messages directly from each quantum as it completes).

Reply:
I see two advantages of doing this inside a job rather than an external service:
- It lets `finalJob` depend on the monitor job, so workflow completion is actually meaningful; you don't have to invent another pathway for the service to inform pilots that their workflow really is done.
- Admins for smaller clusters (i.e. not USDF) don't have to deploy kafka to get an efficient BPS deployment.